### ISSUE TITLE
Regression: Wrapped System.Numerics types are not generating SIMD code

### ISSUE THREAD

**Issue by WysIcoN** [dotnet/coreclr#19438]
Greetings,

I'm running Visual Studio 2017 v15.7.6 with .Net Core 2.1.400-preview-009171. I've noticed that SIMD types from System.Numerics that are wrapped within structs are not generating optimal code. I've tried running from vanilla JIT as well as AOT (both RyuJIT and CPP CodeGen) to no avail.

I've attached a simple project that demonstrates this, along with the disassembly.

Furthermore, it appears that this might be a regression, given it seems to have been resolved in https://github.com/dotnet/coreclr/issues/7508 for a prior version.

It would be awesome if this issue can be addressed ASAP.

[SimdCodeGen.zip](https://github.com/dotnet/coreclr/files/2280547/SimdCodeGen.zip)
[NativeVector_Disassembly.txt](https://github.com/dotnet/coreclr/files/2280548/NativeVector_Disassembly.txt)
[WrappedVector_Disassembly.txt](https://github.com/dotnet/coreclr/files/2280549/WrappedVector_Disassembly.txt)

Thanks.

@CarolEidt @sivarv

**Comment by WysIcoN** [comment]
Additionally, even in the unwrapped case of directly using System.Numerics.Vector4, it seems like there are some odd/unnecessary instructions there, for example vmovapd, which shouldn't be generated since we're operating on floats and not doubles. Here's the equivalent disassembly from C++, which is what I would expect RyuJIT to generate given this is a super trivial example:

	float4 v0 = (float4)(1.0f, 2.0f, 3.0f, 4.0f);
00007FF6775913A3  movaps      xmm0,xmmword ptr [__xmm@40800000408000004080000040800000 (07FF677598BB0h)]  
	float4 v0 = (float4)(1.0f, 2.0f, 3.0f, 4.0f);
00007FF6775913AA  movaps      xmmword ptr [v0],xmm0  
	float4 v1 = (float4)(5.0f, 6.0f, 7.0f, 8.0f);
00007FF6775913AF  movaps      xmm0,xmmword ptr [__xmm@41000000410000004100000041000000 (07FF677598BD0h)]  
00007FF6775913B6  movaps      xmmword ptr [v1],xmm0  
	float4 v2 = (v0 * v1) - (v1 / v0 + v1);
00007FF6775913BB  movaps      xmm0,xmmword ptr [v0]  
00007FF6775913C0  mulps       xmm0,xmmword ptr [v1]  
00007FF6775913C5  movaps      xmm1,xmmword ptr [v1]  
00007FF6775913CA  divps       xmm1,xmmword ptr [v0]  
00007FF6775913CF  addps       xmm1,xmmword ptr [v1]  
00007FF6775913D4  subps       xmm0,xmm1  
00007FF6775913D7  movaps      xmmword ptr [v2],xmm0  
	float4 v3 = (v2 * v0) - (v2 / v0 + v1);
00007FF6775913DC  movaps      xmm0,xmmword ptr [v2]  
00007FF6775913E1  mulps       xmm0,xmmword ptr [v0]  
00007FF6775913E6  movaps      xmm1,xmmword ptr [v2]  
00007FF6775913EB  divps       xmm1,xmmword ptr [v0]  
00007FF6775913F0  addps       xmm1,xmmword ptr [v1]  
00007FF6775913F5  subps       xmm0,xmm1  
00007FF6775913F8  movaps      xmmword ptr [v3],xmm0

**Comment by AndyAyersMS** [comment]
I see identical codegen for `Main` in Core 2.0, Core 2.1, and current master
```
       56                   push     rsi
       4881ECF0000000       sub      rsp, 240
       C5F877               vzeroupper 
       C4E17829B424E0000000 vmovaps  qword ptr [rsp+E0H], xmm6
       C4E17829BC24D0000000 vmovaps  qword ptr [rsp+D0H], xmm7

G_M65024_IG02:
       C4E17A1005D0020000   vmovss   xmm0, dword ptr [reloc @RWD00]
       C4E17A100DCB020000   vmovss   xmm1, dword ptr [reloc @RWD04]
       C4E17A1015C6020000   vmovss   xmm2, dword ptr [reloc @RWD08]
       C4E17A101DC1020000   vmovss   xmm3, dword ptr [reloc @RWD12]
       C4E15857E4           vxorps   xmm4, xmm4
       C4E15A10E3           vmovss   xmm4, xmm4, xmm3
       C4E15973FC04         vpslldq  xmm4, 4
       C4E15A10E2           vmovss   xmm4, xmm4, xmm2
       C4E15973FC04         vpslldq  xmm4, 4
 ... rest omitted ...
```
Can you double-check that you have enabled the jit's optimizer? Make sure you are using a release build of your project.

Additionally, if you are using VS to look at the code, you need to configure VS not to disable jit optimizations, by making sure the highlighted debugger option is NOT checked:

![image](https://user-images.githubusercontent.com/10121823/44180360-cdaaa100-a0b0-11e8-8fd5-ceac3bd1be80.png)

**Comment by WysIcoN** [comment]
Hi Andy,

Thanks much for the quick response and the specific details. The code I posted above was indeed from debug as I didn't expect that debug wouldn't be vectorized in the wrapped case, rather that it would just not be optimized.

However, even in release, there are differences in code generation -- several vmovapd instructions are present in the wrapped case that aren't in the native (unwrapped) case. I've attached the disassembly from release in this case.

Furthermore, this can be seen in terms of performance as well, where the native case runs **faster than 2x** compared to the wrapped case (release builds, of course), which is something I wouldn't expect. For example, on my machine (which is an i7-7820HK @ 2.90 GHz), running those ops for 1M vectors takes about 4ms in the unwrapped case, while it takes about 10ms in the wrapped case.

I've attached the updated program as well, and you can test both paths via commenting out the #define USE_NATIVE above. Please let me know if I'm missing something or if there is anything I can help with further on this.

Thanks for your help.

[NativeVector_Disassembly.txt](https://github.com/dotnet/coreclr/files/2298906/NativeVector_Disassembly.txt)
[WrappedVector_Disassembly.txt](https://github.com/dotnet/coreclr/files/2298907/WrappedVector_Disassembly.txt)
[SimdCodeGen_Perf.zip](https://github.com/dotnet/coreclr/files/2298905/SimdCodeGen_Perf.zip)

**Comment by AndyAyersMS** [comment]
Thanks -- looks like the jit is being overly conservative about forcing temps to the stack.

Perhaps something easily fixed? Eg in `Main`, `V72` ends up on the stack because of the following sequence.
```
fgMorphCopyBlock:block assignment to morph:
               [000444] -----+------              /--*  LCL_VAR   simd16 V25 tmp15        
               [000452] -A------R---              *  ASG       simd16 (copy)
               [000451] x----+-N----              \--*  BLK(16)   simd16
               [000450] -----+------                 \--*  ADDR      byref 
               [000449] D----+-N----                    \--*  LCL_VAR   simd16 V72 tmp62        
 this requires a CopyBlock.

Local V72 should not be enregistered because: written in a block op
Replacing block node [000451] with lclVar V72
GenTreeNode creates assertion:
               [000452] -A------R---              *  ASG       simd16 (copy)
In BB03 New Local Copy     Assertion: V72 == V25 index=#02, mask=0000000000000002

fgMorphTree BB03, stmt 25 (after)
               [000444] -----+------              /--*  LCL_VAR   simd16 V25 tmp15        
               [000452] -A---+--R---              *  ASG       simd16 (copy)
               [000449] D----+-N----              \--*  LCL_VAR   simd16 V72 tmp62
```
@CarolEidt could we call `fgMorphBlockOperand` early on in `fgMorphCopyBlock` or are there complications?

**Comment by CarolEidt** [comment]
> could we call fgMorphBlockOperand early on in fgMorphCopyBlock or are there complications?

One of the things that I did when making the early "first class structs" changes was to ensure that I was generating exactly the same code (i.e. just as conservative as before), however I think the right place to fix this is to avoid create the `BLK` or `OBJ` node on the lhs of an assignment in the case where it's a lclVar. There are comments here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L1322 and here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L1687 where I believe we should avoid creating the block nodes.

There are likely to be complications, i.e. downstream assumptions that would need to be fixed, but I think that avoiding these early is the right way to go. There are probably other places that we introduce these as well.

**Comment by AndyAyersMS** [comment]
This starts out as a field assignment:
```
[000453] ------------     *  STMT      void  
[000444] ------------     |  /--*  LCL_VAR   simd16 V25 tmp15        
[000452] -A------R---     \--*  ASG       simd16 (copy)
[000451] ------------        \--*  BLK(16)   simd16
[000450] ------------           \--*  ADDR      byref 
[000449] ------------              \--*  FIELD     struct v
[000428] L-----------                 \--*  ADDR      byref 
[000429] ------------                    \--*  LCL_VAR   struct V07 loc6         
```
and then the one and only field of V07 gets promoted to give us the tree above.

So should we try and do this simplification when promoting too?

**Comment by AndyAyersMS** [comment]
FWIW with a crude prototype -- simplify BLK(ADDR(LCL)) in the `fgMarkAddrTakenLocalsPostCB` and marking promoted SIMD fields as `lvRegStruct` (which is probably a cheat), things look pretty decent... a few extra reg movs in there but overall not bad.

So it's the conservatism about handling copies from promoted SIMD type fields that causes the codegen you see. Hopefully the work we have in mind for "first class structs" cleans this up more officially.

```asm
       C4E17A1005B5020000   vmovss   xmm0, dword ptr [reloc @RWD00]
       C4E17A100DB0020000   vmovss   xmm1, dword ptr [reloc @RWD04]
       C4E17A1015AB020000   vmovss   xmm2, dword ptr [reloc @RWD08]
       C4E17A101DA6020000   vmovss   xmm3, dword ptr [reloc @RWD12]
       C4E15857E4           vxorps   xmm4, xmm4
       C4E15A10E3           vmovss   xmm4, xmm4, xmm3
       C4E15973FC04         vpslldq  xmm4, 4
       C4E15A10E2           vmovss   xmm4, xmm4, xmm2
       C4E15973FC04         vpslldq  xmm4, 4
       C4E15A10E1           vmovss   xmm4, xmm4, xmm1
       C4E15973FC04         vpslldq  xmm4, 4
       C4E15A10E0           vmovss   xmm4, xmm4, xmm0
       C4E17828C4           vmovaps  xmm0, xmm4
       C4E17A100D71020000   vmovss   xmm1, dword ptr [reloc @RWD16]
       C4E17A10156C020000   vmovss   xmm2, dword ptr [reloc @RWD20]
       C4E17A101D67020000   vmovss   xmm3, dword ptr [reloc @RWD24]
       C4E17A102562020000   vmovss   xmm4, dword ptr [reloc @RWD28]
       C4E15057ED           vxorps   xmm5, xmm5
       C4E15210EC           vmovss   xmm5, xmm5, xmm4
       C4E15173FD04         vpslldq  xmm5, 4
       C4E15210EB           vmovss   xmm5, xmm5, xmm3
       C4E15173FD04         vpslldq  xmm5, 4
       C4E15210EA           vmovss   xmm5, xmm5, xmm2
       C4E15173FD04         vpslldq  xmm5, 4
       C4E15210E9           vmovss   xmm5, xmm5, xmm1
       C4E17828CD           vmovaps  xmm1, xmm5
       C4E17859D1           vmulps   xmm2, xmm0, xmm1
       C4E1705ED8           vdivps   xmm3, xmm1, xmm0
       C4E16058D9           vaddps   xmm3, xmm1
       C4E1685CD3           vsubps   xmm2, xmm3
       C4E16859D8           vmulps   xmm3, xmm2, xmm0
       C4E1685EE0           vdivps   xmm4, xmm2, xmm0
       C4E17828C4           vmovaps  xmm0, xmm4
       C4E17828E0           vmovaps  xmm4, xmm0
       C4E17828C4           vmovaps  xmm0, xmm4
       C4E17858C1           vaddps   xmm0, xmm1
       C4E1605CC8           vsubps   xmm1, xmm3, xmm0
       C4E17828C1           vmovaps  xmm0, xmm1
       C4E17828C8           vmovaps  xmm1, xmm0
       C4E17828C1           vmovaps  xmm0, xmm1
       C4E16859C0           vmulps   xmm0, xmm2, xmm0
```
Test claims time dropped from 13ms to 6ms.

**Comment by tannergooding** [comment]
I don't think that `Vector2`/`Vector3`/`Vector4` have enough existing handling to be treated as "primitive" types. Unlike the `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` types, they were not originally mapped to the `__m64`, `__m128`, and `__m256` ABI types and may have different packing/etc.

However, I do think that these types could be treated as HVA types and that would be effectively what we want here (and would also expand to any user defined structs of the right shape). I think this is tracked by one of the existing "first class struct" issues.

**Comment by mikedn** [comment]
> I don't think that Vector2/Vector3/Vector4 have enough existing handling to be treated as "primitive" types. Unlike the Vector64<T>, Vector128<T>, and Vector256<T> types, they were not originally mapped to the __m64, __m128, and __m256 ABI types and may have different packing/etc.

I don't think you understand what I'm asking. And anyway that was a question for Carol Eidt.

**Comment by CarolEidt** [comment]
> So should we try and do this simplification when promoting too?

Yes, I think so. That's what I meant by "there are probably other places that we introduce these as well". However, I suspect that this one *may* have been introduced by `impAssignStructPtr`, so the first comment I mention would apply. That is, any node that carries its size info (e.g. via a class handle) shouldn't need a BLK node created for it until such time (if any) as it gets transformed to a more general indirection (with a base and offset for example) where it loses the type/size info.

**Comment by AndyAyersMS** [comment]
Makes sense, thanks.

Trying something like this out I run into cases where in promotion we replace a struct that wraps a single non-struct field with the field itself, and this leads to mistyped assignments if the other side of the operation hasn't undergone promotion (eg `[000173]` in the prototype below). 

It looks like morph tries to tolerate some cases mismatched types for struct assignments but currently an assign is only recognized as a copy block op if the destination is a struct.

```
;; CURRENT 

[000175] ------------   *  STMT      void  
[000169] ------------   |  /--*  LCL_VAR   struct V13 tmp3         
[000174] -A------R---   \--*  ASG       struct (copy)
[000173] ------------      \--*  OBJ(8)    struct
[000172] ------------         \--*  ADDR      byref 
[000171] ------------            \--*  FIELD     struct _pointer
[000150] L-----------               \--*  ADDR      byref 
[000151] ------------                  \--*  LCL_VAR   struct V12 tmp2 

Replacing the field in promoted struct with a local var:

[000169] ------------   /--*  LCL_VAR   struct(P) V13 tmp3         
                        /--*    byref  V13._value (offs=0x00) -> V21 tmp11        
[000174] -A------R---   *  ASG       struct (copy)
[000173] ------------   \--*  OBJ(8)    struct
[000172] ------------      \--*  ADDR      byref 
[000171] ------------         \--*  LCL_VAR   byref  V19 tmp9 

;; PROTOTYPE

[000174] ------------   *  STMT      void  
[000169] ------------   |  /--*  LCL_VAR   struct V13 tmp3         
[000173] -A------R---   \--*  ASG       struct (copy)
[000171] ------------      \--*  FIELD     struct _pointer
[000150] L-----------         \--*  ADDR      byref 
[000151] ------------            \--*  LCL_VAR   struct V12 tmp2 

Replacing the field in promoted struct with a local var:

[000169] ------------   /--*  LCL_VAR   struct(P) V13 tmp3         
                        /--*    byref  V13._value (offs=0x00) -> V21 tmp11        
[000173] -A------R---   *  ASG       struct
[000171] D------N----   \--*  LCL_VAR   byref  V19 tmp9 
```

**Comment by CarolEidt** [comment]
> Can't SIMD types be treated as primitive types? Seems counterintuitive that they're structs, they end up in registers and copying a SIMD value reduces to a single instruction.

According to the spec, they are actually structs (value types). Years ago, when I prototyped `Vector<T>` (which I called `SIMDVector<T>`, as I also envisioned a `Vector<T>` that had arbitrary length), I added them as primitive types, but my recollection is that doing so caused as many problems as it solved, especially given that they are fundamentally structs (value types) according to the vm/type system. That said, I don't think there's any reason that the JIT couldn't treat them as fundamental types, and in fact, in some sense the fact that they are typed differently from `TYP_STRUCT` was intended to take us in that direction. Ultimately, I expect/hope that most of the places we pessimize the handling of SIMD types over primitive types will be addressed by changing `varTypeIsStruct(type)` to `type == TYP_STRUCT`.

**Comment by CarolEidt** [comment]
> currently an assign is only recognized as a copy block op if the destination is a struct.

It seems to me that, in the example you give, it would make sense for the assignment to become a non-struct assignment, as it is now simply an assignment of the individual field. But it seems that doing so would require additional smarts in the code that replaces promoted structs with lclVars.

**Comment by AndyAyersMS** [comment]
Agree -- for instance today we won't replace a non-field access of a single-field promoted struct with the field.

Not sure the best way to engineer this either. It seems clunky for struct assign leaves to independently decide they should update to promoted fields. Likely the parent assign needs to look at both its operands and determine whether the assignment should be expressed as a field assignment or left as a struct assignment. But putting this logic into promotion update duplicates the logic from block copy.

So I'd say we should add this case to the first class structs example set and revisit as that work progresses.

**Comment by CarolEidt** [comment]
@AndyAyersMS - Previously I missed your comment where you talk about `lvRegStruct`. I don't think that setting that is a cheat at all. Basically it just means that this local is a struct that fits in a register and is not being promoted (i.e. in this case, not itself being promoted into sub-fields, which we sometimes do for `VectorN` when the individual fields are accessed more often than the full struct). Anyway, I've incorporated that change into a larger WIP that eliminates some of the other struct pessimisms, and it gets the expected code.
I still have some issues to address with value numbering in that WIP, so I might consider pulling out just the `lvRegStruct` change.

**Comment by CarolEidt** [comment]
I've incorporated this test case into my PR #22255 which I will merge as soon as the tests complete.
Here's the first block of the loop:
Before:
```
G_M31315_IG05:
       vmovupd  xmm0, xmmword ptr [reloc @RWD00]
       vmovapd  xmmword ptr [rbp-50H], xmm0
       vmovupd  xmm0, xmmword ptr [reloc @RWD16]
       vmovapd  xmmword ptr [rbp-60H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-50H]
       vmovapd  xmm1, xmmword ptr [rbp-60H]
       vmulps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-70H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-70H]
       vmovapd  xmm1, xmmword ptr [rbp-60H]
       vmovapd  xmm2, xmmword ptr [rbp-50H]
       vdivps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-80H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-80H]
       vmovapd  xmm2, xmmword ptr [rbp-60H]
       vaddps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-90H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-90H]
       vsubps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-A0H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-A0H]
       vmovapd  xmm1, xmmword ptr [rbp-50H]
       vmulps   xmm1, xmm0, xmm1
       vmovapd  xmmword ptr [rbp-B0H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-B0H]
       vmovapd  xmm2, xmmword ptr [rbp-50H]
       vdivps   xmm3, xmm0, xmm2
       vmovapd  xmmword ptr [rbp-C0H], xmm3
       vmovapd  xmm2, xmmword ptr [rbp-C0H]
       vmovapd  xmm3, xmmword ptr [rbp-60H]
       vaddps   xmm2, xmm3
       vmovapd  xmmword ptr [rbp-D0H], xmm2
       vmovapd  xmm2, xmmword ptr [rbp-D0H]
       vsubps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-E0H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-E0H]
       vmulps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-F0H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-F0H]
       inc      dword ptr [rbx+20]
       mov      rcx, gword ptr [rbx+8]
       mov      edx, dword ptr [rbx+16]
       cmp      dword ptr [rcx+8], edx
       jbe      SHORT G_M31315_IG06
       lea      eax, [rdx+1]
       mov      dword ptr [rbx+16], eax
       movsxd   rdx, edx
       shl      rdx, 4
       vmovupd  xmmword ptr [rcx+rdx+16], xmm0
       jmp      SHORT G_M31315_IG07
```
After:
```
G_M31315_IG05:
       vmovupd  xmm0, xmmword ptr [reloc @RWD00]
       vmovupd  xmm1, xmmword ptr [reloc @RWD16]
       vmulps   xmm2, xmm0, xmm1
       vdivps   xmm3, xmm1, xmm0
       vaddps   xmm3, xmm1
       vsubps   xmm2, xmm3
       vmulps   xmm3, xmm2, xmm0
       vdivps   xmm4, xmm2, xmm0
       vaddps   xmm1, xmm4, xmm1
       vsubps   xmm3, xmm1
       vmulps   xmm2, xmm3
       inc      dword ptr [rbx+20]
       mov      rcx, gword ptr [rbx+8]
       mov      edx, dword ptr [rbx+16]
       cmp      dword ptr [rcx+8], edx
       jbe      SHORT G_M31315_IG06
       lea      eax, [rdx+1]
       mov      dword ptr [rbx+16], eax
       movsxd   rdx, edx
       shl      rdx, 4
       vmovupd  xmmword ptr [rcx+rdx+16], xmm2
       jmp      SHORT G_M31315_IG07

```

**Comment by WysIcoN** [comment]
Awesome! Thanks for the update and fixing this Carol.
Ani


Sent from Yahoo Mail for iPhone


On Tuesday, March 26, 2019, 9:04 AM, Carol Eidt <notifications@github.com> wrote:


I've incorporated this test case into my PR #22255 which I will merge as soon as the tests complete.
Here's the first block of the loop:
Before:
G_M31315_IG05:
       vmovupd  xmm0, xmmword ptr [reloc @RWD00]
       vmovapd  xmmword ptr [rbp-50H], xmm0
       vmovupd  xmm0, xmmword ptr [reloc @RWD16]
       vmovapd  xmmword ptr [rbp-60H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-50H]
       vmovapd  xmm1, xmmword ptr [rbp-60H]
       vmulps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-70H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-70H]
       vmovapd  xmm1, xmmword ptr [rbp-60H]
       vmovapd  xmm2, xmmword ptr [rbp-50H]
       vdivps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-80H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-80H]
       vmovapd  xmm2, xmmword ptr [rbp-60H]
       vaddps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-90H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-90H]
       vsubps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-A0H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-A0H]
       vmovapd  xmm1, xmmword ptr [rbp-50H]
       vmulps   xmm1, xmm0, xmm1
       vmovapd  xmmword ptr [rbp-B0H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-B0H]
       vmovapd  xmm2, xmmword ptr [rbp-50H]
       vdivps   xmm3, xmm0, xmm2
       vmovapd  xmmword ptr [rbp-C0H], xmm3
       vmovapd  xmm2, xmmword ptr [rbp-C0H]
       vmovapd  xmm3, xmmword ptr [rbp-60H]
       vaddps   xmm2, xmm3
       vmovapd  xmmword ptr [rbp-D0H], xmm2
       vmovapd  xmm2, xmmword ptr [rbp-D0H]
       vsubps   xmm1, xmm2
       vmovapd  xmmword ptr [rbp-E0H], xmm1
       vmovapd  xmm1, xmmword ptr [rbp-E0H]
       vmulps   xmm0, xmm1
       vmovapd  xmmword ptr [rbp-F0H], xmm0
       vmovapd  xmm0, xmmword ptr [rbp-F0H]
       inc      dword ptr [rbx+20]
       mov      rcx, gword ptr [rbx+8]
       mov      edx, dword ptr [rbx+16]
       cmp      dword ptr [rcx+8], edx
       jbe      SHORT G_M31315_IG06
       lea      eax, [rdx+1]
       mov      dword ptr [rbx+16], eax
       movsxd   rdx, edx
       shl      rdx, 4
       vmovupd  xmmword ptr [rcx+rdx+16], xmm0
       jmp      SHORT G_M31315_IG07

After:
G_M31315_IG05:
       vmovupd  xmm0, xmmword ptr [reloc @RWD00]
       vmovupd  xmm1, xmmword ptr [reloc @RWD16]
       vmulps   xmm2, xmm0, xmm1
       vdivps   xmm3, xmm1, xmm0
       vaddps   xmm3, xmm1
       vsubps   xmm2, xmm3
       vmulps   xmm3, xmm2, xmm0
       vdivps   xmm4, xmm2, xmm0
       vaddps   xmm1, xmm4, xmm1
       vsubps   xmm3, xmm1
       vmulps   xmm2, xmm3
       inc      dword ptr [rbx+20]
       mov      rcx, gword ptr [rbx+8]
       mov      edx, dword ptr [rbx+16]
       cmp      dword ptr [rcx+8], edx
       jbe      SHORT G_M31315_IG06
       lea      eax, [rdx+1]
       mov      dword ptr [rbx+16], eax
       movsxd   rdx, edx
       shl      rdx, 4
       vmovupd  xmmword ptr [rcx+rdx+16], xmm2
       jmp      SHORT G_M31315_IG07

**Comment by CarolEidt** [comment]
Fixed with #22255
